### PR TITLE
Add icg strstr

### PR DESCRIPTION
**Pull request by GiulioLyons**
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

### CODE REVIEW

File: libr/core/cmd_info.c [hunk 1]
```diff
@@ -1255,6 +1255,19 @@ static int cmd_info(void *data, const char *input) {
...
 				}
 				bool fullGraph = true;
+				if (input[3] != ' ') {
+					bool fullGraph = false;
+					r_list_foreach (obj->classes, iter, cls) {
+					        if (cls->super && strstr (cls->super, input + 3)) {
+							r_cons_printf ("agn %s\n", cls->super);
+							r_cons_printf ("agn %s\n", cls->name);
+							r_cons_printf ("age %s %s\n", cls->super, cls->name);
+						} else if (strstr (cls->name, input + 3)){
```
Comment (trufae) [inline libr/core/cmd_info.c:1265]:
```suggestion
						} else if (strstr (cls->name, input + 3)) {
```

File: libr/core/cmd_info.c [hunk 2]
```diff
@@ -1255,6 +1255,19 @@ static int cmd_info(void *data, const char *input) {
 					break;
 				}
 				bool fullGraph = true;
+				if (input[3] != ' ') {
+					bool fullGraph = false;
+					r_list_foreach (obj->classes, iter, cls) {
+					        if (cls->super && strstr (cls->super, input + 3)) {
```
Comment (trufae) [inline libr/core/cmd_info.c:1261]: keep that input+3 pointer in a variable outside the loop

**Comment by GiulioLyons** [comment]
Its working now, can use icg with or without a string passed as arg

**Comment by trufae** [comment]
Please check and solve the LGTM warnings `8 for Wrong type of arguments to formatting function` https://lgtm.com/projects/g/radareorg/radare2/rev/pr-ea09698a3fc588731e0d2b38a98b3125db8a0e08

File: libr/core/cmd_info.c [hunk 3]
```diff
@@ -1255,6 +1255,19 @@ static int cmd_info(void *data, const char *input) {
 					break;
 				}
 				bool fullGraph = true;
+				if (input[2] == ' ' && strcmp(input + 3, "")) {
```
Comment (trufae) [inline libr/core/cmd_info.c:1258]: what's exactly what you wanna check here? if there's a space and not a null byte after it?
```suggestion
				if (input[2] == ' ' && input[3]) {
```

then this code is simpler and more readable, but what if there are many spaces? you shoul dbetter do something like this:

```suggestion
				const char *match = r_str_trim_head_ro (input + 2);
				if (*match) {
```

File: libr/core/cmd_info.c [hunk 4]
```diff
@@ -1255,6 +1255,19 @@ static int cmd_info(void *data, const char *input) {
...
+					const char *match = r_str_trim_head_ro (input + 3);
```
Comment (trufae) [inline libr/core/cmd_info.c:1259]:
```suggestion
```

File: libr/core/cmd_info.c [hunk 5]
```diff
@@ -1255,6 +1255,19 @@ static int cmd_info(void *data, const char *input) {
...
+					r_list_foreach (obj->classes, iter, cls) {
+					        if (cls->super && strstr (cls->super, match)) {
+							r_cons_printf ("agn %s\n", cls->super);
```
Comment (trufae) [inline libr/core/cmd_info.c:1262]: this can introduce code injection vulnerabilities. quote the whole command and filter the cls->super value. maybe we can have a filtering api to do that automatically

File: libr/core/cmd_info.c [hunk 6]
```diff
@@ -1255,6 +1255,19 @@ static int cmd_info(void *data, const char *input) {
...
+							r_cons_printf ("agn %s\n", cls->name);
+							r_cons_printf ("age %s %s\n", cls->super, cls->name);
```
Comment (trufae) [inline libr/core/cmd_info.c:1264]: and here

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/core/cmd_info.c [hunk 7]
```diff
@@ -1255,6 +1255,20 @@ static int cmd_info(void *data, const char *input) {
 					break;
 				}
 				bool fullGraph = true;
+				const char *match = r_str_trim_head_ro (input + 2);
+				if (*match) {
+					const char *match = r_str_trim_head_ro (input + 3);
```
Comment (trufae) [inline libr/core/cmd_info.c:1260]:
```suggestion
```

**Comment by trufae** [comment]
and also maybe interesting to cover this [strstr] for the other `ic` commands.

also, im missing a `ic,` command that prints in table format, would you like to add it too? (in another pr)

**Comment by trufae** [comment]
dont forget to git pull because i have pushed changes

File: libr/core/cmd_info.c [hunk 8]
```diff
@@ -16,7 +16,7 @@ static const char *help_msg_i[] = {
 	"ib", "", "reload the current buffer for setting of the bin (use once only)",
 	"ic", "", "List classes, methods and fields",
 	"icc", "", "List classes, methods and fields in Header Format",
-	"icg", "", "List classes as agn/age commands to create class hirearchy graphs",
+	"icg", " [strstr]", "List classes as agn/age commands to create class hirearchy graphs",
```
Comment (trufae) [inline libr/core/cmd_info.c:19]:
```suggestion
	"icg", " [str]", "List classes as agn/age commands to create class hirearchy graphs (matches str if provided)",
```

**Comment by trufae** [comment]
fix the LGTM issues. the super->extends thing shouldnt happen before 5.8

Screenshot 2022-06-08 at 23 42 01

**Comment by GiulioLyons** [comment]
> fix the LGTM issues. the super->extends thing shouldnt happen before 5.8
> 
> Screenshot 2022-06-08 at 23 42 01

I think these errors are gone now?

File: libr/core/cmd_info.c [hunk 9]
```diff
@@ -1255,7 +1255,19 @@ static int cmd_info(void *data, const char *input) {
...
+					    if (cls->super && strstr (cls->super, match)) {
+							r_cons_printf ("agn %s\n", cls->super);
+							r_cons_printf ("agn %s\n", cls->name);
+							r_cons_printf ("age %s %s\n", cls->super, cls->name);
+						} else if (strstr (cls->name, match)) {
+							r_cons_printf ("agn %s\n", cls->name);
+						}
+					}
+				}
+                else if (fullGraph) {
```
[the remark this answers is not in the document]
Comment (GiulioLyons) [reply libr/core/cmd_info.c]: Should I remove or comment out the code below that? Cause full graph gets set to true it never gets hit

File: test/db/cmd/cmd_i [hunk 10]
```diff
@@ -509,6 +509,39 @@ EXPECT=<<EOF
 EOF
 RUN
 
+NAME=icg
+FILE=bins/elf/libmagic.so
+CMDS=<<EOF
+icg
```
[the remark this answers is not in the document]
Comment (GiulioLyons) [reply test/db/cmd/cmd_i:515]: I haven't done the rest of the ic commands yet I can do them in this pr if that's ok

**Comment by trufae** [comment]
There was a ticket for this thing? if so please refer to it. now let's wait for the CI and merge if it passes the test. but you shuold add ?e between each agn line and the output of the test should have empty lines to be valid

**Comment by GiulioLyons** [comment]
I added the spaces in the test and heres the issue #14686
